relayer/workers/parachaincommitmentrelayer/beefy-listener.go

```diff
@@ -75,81 +93,122 @@ func (li *BeefyListener) onDone(ctx context.Context) error {
 }
 
 func (li *BeefyListener) subBeefyJustifications(ctx context.Context) error {
-	ch := make(chan interface{})
+	headers := make(chan *gethTypes.Header, 5)
 
-	li.log.Info("Subscribing to relay chain light client for new mmr payloads")
-	sub, err := li.relaychainConn.GetAPI().Client.Subscribe(context.Background(), "beefy", "subscribeJustifications", "unsubscribeJustifications", "justifications", ch)
-	if err != nil {
-		panic(err)
-	}
-	defer sub.Unsubscribe()
+	li.ethereumConn.GetClient().SubscribeNewHead(ctx, headers)
 
 	for {
 		select {
 		case <-ctx.Done():
 			return li.onDone(ctx)
-		case msg := <-ch:
+		case gethheader := <-headers:
+			// Query LightClientBridge contract's ContractNewMMRRoot events
+			blockNumber := gethheader.Number.Uint64()
+			var beefyLightClientEvents []*beefylightclient.ContractNewMMRRoot
 
-			signedCommitment := &store.SignedCommitment{}
-			err := types.DecodeFromHexString(msg.(string), signedCommitment)
+			contractEvents, err := li.queryBeefyLightClientEvents(ctx, blockNumber, &blockNumber)
 			if err != nil {
-				li.log.WithError(err).Error("Failed to decode BEEFY commitment messages")
+				li.log.WithError(err).Error("Failure fetching event logs")
+				return err
 			}
+			beefyLightClientEvents = append(beefyLightClientEvents, contractEvents...)
+
+			if len(beefyLightClientEvents) > 0 {
+				li.log.Info(fmt.Sprintf("Found %d BeefyLightClient ContractNewMMRRoot events on block %d", len(beefyLightClientEvents), blockNumber))
+			}
+			li.processBeefyLightClientEvents(ctx, beefyLightClientEvents)
+		}
+	}
+}
+
+// processLightClientEvents matches events to BEEFY commitment info by transaction hash
+func (li *BeefyListener) processBeefyLightClientEvents(ctx context.Context, events []*beefylightclient.ContractNewMMRRoot) error {
+	for _, event := range events {
+
+		contractAbi, err := abi.JSON(strings.NewReader(string(beefylightclient.ContractABI)))
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		eventUnpacked, err := contractAbi.Unpack("FinalVerificationSuccessful", event.Raw.Data)
+		if err != nil {
+			return err
+		}
+
+		relayChainBlockNumber := (eventUnpacked[1].(*big.Int)).Int64()
+
+		li.log.WithFields(logrus.Fields{
+			"relayChainBlockNumber": relayChainBlockNumber,
+			"ethereumBlockNumber":   event.Raw.BlockNumber,
+			"ethereumTxHash":        event.Raw.TxHash.Hex(),
+		}).Info("Witnessed a new MMRRoot event")
 
-			blockNumber := signedCommitment.Commitment.BlockNumber
+		li.log.WithField("blockNumber", relayChainBlockNumber).Info("Getting hash for relay chain block")
+		blockHash, err := li.relaychainConn.GetAPI().RPC.Chain.GetBlockHash(uint64(relayChainBlockNumber))
+		if err != nil {
+			li.log.WithError(err).Error("Failed to get block hash")
+			return err
+		}
+		li.log.WithField("blockHash", blockHash.Hex()).Info("Got blockhash")
+
+		// TODO this just queries the latest MMR leaf in the latest MMR and our latest parahead from the relaychain.
```

Comment:
Let's separate this logic into it's own function, e. g. `submitMessages(blockHash)`

Reply:
will handle as part of the catchup/graceful recovery PR, as a bunch will be refactored there anyway